src/components/reports/views.tsx

```diff
@@ -0,0 +1,461 @@
+import moment from 'moment';
+import React, { ReactElement } from 'react';
+import { DATE } from '../../layouts';
+import { RouteLinker } from '../app/context';
+
+import { IOrganizationQuota, IV3OrganizationResource } from '../../lib/cf/types';
+import { IBillableByOrganisationAndService, IBillableByOrganisationAndSpaceAndService, IBillableByService } from './controllers';
+
+interface IOrganizationsReportProperties {
+  readonly organizations: readonly IV3OrganizationResource[];
+  readonly trialOrgs: readonly IV3OrganizationResource[];
+  readonly billableOrgs: readonly IV3OrganizationResource[];
+  readonly linkTo: RouteLinker;
+  readonly orgQuotaMapping: {[key: string]: IOrganizationQuota};
+  readonly orgTrialExpirys: {[key: string]: Date};
+}
+
+interface ICostable {
+  readonly incVAT: number;
+  readonly exVAT: number;
+}
+
+interface IOrgCostRecord extends ICostable {
+  readonly orgGUID: string;
+  readonly orgName: string;
+  readonly quotaGUID: string;
+  readonly quotaName: string;
+}
+
+interface IQuotaCostRecord extends ICostable {
+  readonly quotaGUID: string;
+  readonly quotaName: string;
+}
+
+interface ICostReportProperties {
+  readonly date: string;
+  readonly billableEventCount: number;
+  readonly totalBillables: ICostable;
+  readonly orgCostRecords: readonly IOrgCostRecord[];
+  readonly quotaCostRecords: readonly IQuotaCostRecord[];
+}
+
+interface ICostByServiceReport {
+  readonly date: string;
+  readonly billablesByService: readonly IBillableByService[];
+  readonly billablesByOrganisationAndService: readonly IBillableByOrganisationAndService[];
+  readonly billablesByOrganisationAndSpaceAndService: readonly IBillableByOrganisationAndSpaceAndService[];
+}
+interface ID3SankeyNode {
+  name: string;
+}
+
+interface ID3SankeyLink {
+  source: number;
+  target: number;
+  value: number;
+}
+
+interface ID3SankeyInput {
+  nodes: ReadonlyArray<ID3SankeyNode>;
+  links: ReadonlyArray<ID3SankeyLink>;
+}
+
+interface IVisualisationPage {
+  readonly date: string;
+  readonly data?: ID3SankeyInput;
+}
+
+interface ISankeyInputProperties {
+  readonly data: ID3SankeyInput;
+}
+
+class SankeyInput extends React.Component {
+  constructor(public props: ISankeyInputProperties) {
+    super(props);
+  }
+
+  public render(): ReactElement {
+    return this.SankeyInputJSON();
+  }
+
+  private SankeyInputJSON(): ReactElement {
+    return (
+      <script id="data" type="application/json" dangerouslySetInnerHTML={{
+        __html: JSON.stringify(this.props.data),
```

Comment:
This was probably broken before as well, but after a discussion with Rafal we spotted that there's an XSS vulnerability here.

`data` includes organisation "owners", which are based on the organisation metadata which is writeable by org managers. Someone could set the metadata to something like `</script><script>alert('haha pwned')</script>`. We'd be saved by our CSP in modern browsers, but still.

We could avoid the `dangerouslySetInnerHTML` entirely if we base64 encoded the JSON, but we'd need to teach the client side JS to base64 decode it again.

Alternatively we could escape the `/` in `</script>` which I think is a sufficient mitigation.

Reply:
added `.replace('/', '\/')` for good measure...